kratos/sources/model_part.cpp

```diff
@@ -13,21 +13,109 @@
 
 // System includes
 #include <sstream>
+#include <type_traits>
 
 // External includes
 
 // Project includes
 #include "includes/define.h"
 #include "includes/model_part.h"
 #include "includes/exception.h"
-#include "utilities/parallel_utilities.h"
+#include "utilities/reduction_utilities.h"
 
 namespace Kratos
 {
 
 KRATOS_CREATE_LOCAL_FLAG(ModelPart, ALL_ENTITIES, 0);
 KRATOS_CREATE_LOCAL_FLAG(ModelPart, OVERWRITE_ENTITIES, 1);
 
+namespace ModelPartHelperUtilities
+{
+
+template<class TContainerGetter>
+void AddEntitiesFromIds(
+    TContainerGetter&& rContainerGetter,
+    ModelPart* pModelPart,
+    const std::vector<IndexType>& rEntityIds)
+{
+    KRATOS_TRY
+
+    using container_type = std::remove_pointer_t<decltype(std::declval<TContainerGetter>()(std::declval<ModelPart*>()))>;
+
+    if(pModelPart->IsSubModelPart()) { //does nothing if we are on the top model part
+        //obtain from the root model part the corresponding list of nodes
+        ModelPart* root_model_part = &pModelPart->GetRootModelPart();
+
+        // we first sort and unique the given entity ids.
+        std::vector<IndexType> unique_sorted_ids;
+        unique_sorted_ids.resize(rEntityIds.size());
+        IndexPartition<IndexType>(rEntityIds.size()).for_each([&unique_sorted_ids,  &rEntityIds](const auto Index) {
+            unique_sorted_ids[Index] = rEntityIds[Index];
+        });
+        std::sort(unique_sorted_ids.begin(), unique_sorted_ids.end());
+        auto new_last = std::unique(unique_sorted_ids.begin(), unique_sorted_ids.end());
+
+        const auto& r_container = *rContainerGetter(root_model_part);
+
+        // aux is created to avoid doing sort and unique for all the parent model parts and current model part
+        // when insertion is done.
+        container_type aux;
+        aux.reserve(std::distance(unique_sorted_ids.begin(), new_last));
+        for (auto it_id = unique_sorted_ids.begin(); it_id != new_last; ++it_id) {
+            auto it = r_container.find(*it_id);
+
+            KRATOS_ERROR_IF(it == r_container.end())
+                << "while adding " << ModelPart::Container<container_type>::GetEntityName() << "s to submodelpart "
+                << pModelPart->FullName() << ", the "
+                << ModelPart::Container<container_type>::GetEntityName()
+                << " with Id " << *it_id << " does not exist in the root model part";
+
+            // here the hint is valid, hence this will be a simple push_back within the PVS.
```

Comment:
oh, you mean that under the hood it will eb a push_back. Ok

Reply:
exactly :)